zerver/lib/rate_limiter.py

```diff
@@ -569,8 +569,12 @@ def client_is_exempt_from_rate_limiting(request: HttpRequest) -> bool:
 
     # Don't rate limit requests from Django that come from our own servers,
     # and don't rate-limit dev instances
-    client = RequestNotes.get_notes(request).client
-    return (client is not None and client.name.lower() == "internal") and (
+    notes = RequestNotes.get_notes(request)
+    # Take client_name before the Client object's name to ensure edge cases like overriding client=internal in
+    # request bodies work as expected.
+    client_name = notes.client_name or (notes.client is not None and notes.client.name) or ""
```

Comment:
`notes.client_name` is `Optional[str]`, so this use is ambiguous about whether the intention is to check for `None` or `""`. (This is [mentioned](https://zulip.readthedocs.io/en/latest/testing/mypy.html#use-optional-and-none-correctly) in our mypy best practices documentation.)

Reply:
This is really hard to read. Could this be done more simply with having earlier in this function:

* `if client is None: return False` with a comment noting that this can occur if the middleware cannot parse the client name?

Reply:
No: `client` can be `None` while `notes.client_name` is populated, and `client_name` must take priority as noted in the comment to allow POST/PATCH/GET args to override the auto-detection (that auto-detection was a cause of rate limiting being applied to one of the sub-requests in `/emails/generate`).

I can refactor (and disambiguate) this to reduce terseness/aid readability, but removing the `client is not None and` arm of the return is a core part of this fix, no matter where in the function it lives.